tlslite/utils/python_tripledes.py

```diff
@@ -0,0 +1,497 @@
+
+                # Find the permutation value
+                v = Des.__sbox[j][(m << 4) + n]
+
+                # Turn value into bits, add it to result: bn
+                bn[pos] = (v & 8) >> 3
+                bn[pos + 1] = (v & 4) >> 2
+                bn[pos + 2] = (v & 2) >> 1
+                bn[pos + 3] = v & 1
+
+                pos += 4
+
+            # Permutate the concatination of b[1] to b[8] (bn)
+            self.r = self.__permutate(Des.__p, bn)
+
+            # Xor with l[i - 1]
+            self.r = [x ^ y for x, y in zip(self.r, self.l)]
+            self.l = temp_r
+
+            iteration += iteration_adjustment
+
+        # Final permutation of r[16]l[16]
+        self.final = self.__permutate(Des.__fp, self.r + self.l)
+        return self.final
+
+    # Data to be encrypted/decrypted
+    def crypt(self, data, crypt_type):
+        """Crypt the data in blocks, running it through des_crypt()"""
+
+        iv = self.__string_to_bitlist(self.iv)
+
+        # Split the data into blocks, crypting each one seperately
+        i = 0
+        result = []
+        while i < len(data):
+            # Test code for caching encryption results
+            block = self.__string_to_bitlist(data[i:i+8])
+
+            # Xor with iv if using CBC mode
+            if crypt_type == Des.ENCRYPT:
+                block = [x ^ y for x, y in zip(block, iv)]
+
+            processed_block = self.__des_crypt(block, crypt_type)
+
+            if crypt_type == Des.DECRYPT:
+                processed_block = [x ^ y for x, y in zip(processed_block, iv)]
+                iv = block
+            else:
+                iv = processed_block
+
+            # Add the resulting crypted block to our list
+            result.append(self.__bitlist_to_string(processed_block))
+            i += 8
+
+        # Return the full crypted string
+        return b''.join(result)
+
+#############################################
+#               Triple DES                  #
+#############################################
+
+
+class Python_TripleDES(_baseDes):
+    """Triple DES encryption/decrytpion class
+
+    This algorithm uses the DES-EDE3 (when a 24 byte key is supplied) or
+    the DES-EDE2 (when a 16 byte key is supplied) encryption methods.
+    Supports CBC (Cypher Block Chaining) mode.
+    """
+    def __init__(self, key, iv=None):
+        super(Python_TripleDES, self).__init__(iv)
+
+        # Will set crypting key for this object. Either 16/24 bytes long.
+        self.key_size = len(key)
+        if self.key_size not in (16, 24):
+            raise ValueError("Invalid triple DES key size. "
+                             "Key must be either 16 or 24 bytes long")
+        key = self._guard_against_unicode(key)
+
+        self.__key1 = Des(key[:8], self._iv)
+        self.__key2 = Des(key[8:16], self._iv)
+        if self.key_size == 16:
+            self.__key3 = self.__key1
```

Comment:
this will use the same object for key3 as it does for key1, including its state

do we have a test case with a 2-key 3DES that encrypts multiple blocks to verify that IV handling in this case is correct?

Reply:
test_3des_16B_key() method in test cases managing this option